File: answers/13.cross_validation/cross_validation.jl
# K-Fold Cross-Validation in Julia
using Random, Statistics, GLM, DataFrames

println("=== K-Fold Cross-Validation ===\n")
Random.seed!(42)

# Generate synthetic data
n = 100
X = randn(n, 3)
y = 2*X[:,1] + 3*X[:,2] - X[:,3] + randn(n)*0.5

# Manual K-fold implementation
K = 5
fold_ids = shuffle(repeat(1:K, outer=Int(ceil(n/K)))[1:n])

println("--- Manual $K-Fold Cross-Validation ---\n")
mse_scores = Float64[]

for k in 1:K
    # Split data
    train_idx = fold_ids .!= k
    test_idx = fold_ids .== k

    X_train, X_test = X[train_idx, :], X[test_idx, :]
    y_train, y_test = y[train_idx], y[test_idx]

    # Fit model
    df_train = DataFrame(X_train, :auto)
    df_train.y = y_train
    model = lm(@formula(y ~ x1 + x2 + x3), df_train)

    # Predict and evaluate
    df_test = DataFrame(X_test, :auto)
    y_pred = predict(model, df_test)
    mse = mean((y_test .- y_pred).^2)
    push!(mse_scores, mse)

    println("Fold $k: MSE = $(round(mse, digits=4)), n_train = $(sum(train_idx)), n_test = $(sum(test_idx))")
end

println("\nMean CV MSE: $(round(mean(mse_scores), digits=4)) ± $(round(std(mse_scores), digits=4))")

File: answers/13.cross_validation/cross_validation.py
"""
K-Fold Cross-Validation
Assessing model performance using training/testing splits
"""
import numpy as np
from sklearn.model_selection import KFold, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

print("=== K-Fold Cross-Validation ===\n")

# Generate synthetic data
np.random.seed(42)
n = 100
X = np.random.randn(n, 3)
y = 2*X[:, 0] + 3*X[:, 1] - X[:, 2] + np.random.randn(n)*0.5

# Manual K-fold implementation
K = 5
fold_ids = np.random.permutation(np.repeat(range(K), len(y)//K + 1)[:len(y)])

print(f"--- Manual {K}-Fold Cross-Validation ---\n")
mse_scores = []

for k in range(K):
    # Split data
    train_idx = fold_ids != k
    test_idx = fold_ids == k

    X_train, X_test = X[train_idx], X[test_idx]
    y_train, y_test = y[train_idx], y[test_idx]

    # Fit model
    model = LinearRegression()
    model.fit(X_train, y_train)

    # Predict and evaluate
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mse_scores.append(mse)

    print(f"Fold {k+1}: MSE = {mse:.4f}, n_train = {len(y_train)}, n_test = {len(y_test)}")

print(f"\nMean CV MSE: {np.mean(mse_scores):.4f} ± {np.std(mse_scores):.4f}")

# Using sklearn
print(f"\n--- Using sklearn KFold ---\n")
model = LinearRegression()
kf = KFold(n_splits=K, shuffle=True, random_state=42)
cv_scores = -cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')

print(f"CV MSE scores: {cv_scores}")
print(f"Mean: {np.mean(cv_scores):.4f} ± {np.std(cv_scores):.4f}")
